README

DBD::Pg  --  the DBI PostgreSQL interface for Perl

# $Id: README 11578 2008-07-23 17:17:05Z turnstep $

DESCRIPTION:
------------

This is version 2.8.8 of DBD::Pg.  The web site for this interface, and 
the latest version, can be found at:

	http://search.cpan.org/dist/DBD-Pg/

The mailing list is at:

    http://www.nntp.perl.org/group/perl.dbd.pg/

The development of DBD::Pg can be tracked at:

    http://svn.perl.org/modules/DBD-Pg/trunk/

For information about PostgreSQL, visit:

	http://www.postgresql.org/

For information on what has changed for each version, see the Changes files.


IF YOU HAVE PROBLEMS OR COMMENTS:
---------------------------------

Please send any problems and comments to 
<dbd-pg@perl.org>

Please include what OS you are using, and the version of Perl, 
DBI, and DBD::Pg you are using. Also tell which version of 
PostgreSQL DBD::Pg was compiled against, and which version you 
are connecting to. The easiest way to gather all of this 
information is to run "make test", which outputs it all early in 
the tests.

You can also try the #postgresql channel on irc.freenode.net, which 
usually (but not always) has people who can help you with DBD::Pg.

BUG REPORTS
-----------

If you feel certain you have found a bug, you can file a bug report by visiting:
http://rt.cpan.org/Public/Dist/Display.html?Name=DBD-Pg
and selecting the "Report a new bug" link. Please check that the bug 
has not already been reported first.


REQUIREMENTS:
-------------

	build, test, and install Perl 5         (at least 5.6.1)
	build, test, and install the DBI module (at least 1.52)
	build, test, and install PostgreSQL     (at least 7.4)
	build, test, and install Test::Simple   (at least 0.47)

You must also have the pg_config executable installed (to check, type 
"which pg_config" on unix-like systems). If pg_config is not available, 
then you need to install the development package for PostgreSQL. For example 
on Debian: apt-get install postgresql-dev; on RedHat: yum install postgresql-devel. 
This development package is needed even if you already have PostgreSQL up 
and running since DBD::Pg uses it for its installation.

INSTALLATION:
-------------

Before installing, please use the "cpansign -v" program to cryptographically 
verify that your copy of DBD::Pg is complete and valid. The program 
"cpansign" is part of Module::Signature, available from CPAN.

By default Makefile.PL uses App::Info to find the location of the
PostgreSQL library and include directories.  However, if you want to
control it yourself, define the environment variables POSTGRES_INCLUDE 
and POSTGRES_LIB, or define just POSTGRES_HOME. Note that if you have 
compiled PostgreSQL with SSL support, you must define the POSTGRES_LIB
environment variable and add "-lssl" to it, like this:

	export POSTGRES_LIB="/usr/local/pgsql/lib -lssl"

The usual steps to install DBD::Pg:

	1.   perl Makefile.PL
	2.   make
	3.   make test
	4.   make install

Do steps 1 to 3 as a normal user, not as root!

If the script cannot find the pg_config information itself, it will 
ask you for the path to it. Enter the complete path to the pg_config 
file here, including the name of the file itself.

TESTING:
--------

The tests rely on being able to connect to a valid Postgres database. 
The easiest way to ensure this is to set the following environment variables:

	DBI_DSN=dbi:Pg:dbname=<database>
	DBI_USER=<username>
	DBI_PASS=<password>

If you are running on a non-standard port, you must set PGPORT or 
add the port to the DBI_DSN variable like this:

	DBI_DSN='dbi:Pg:dbname=<database>;port=<port#>'

Put double quotes around the dbname if it has a semicolon 
or a space inside of it:

	DBI_DSN='dbi:Pg:dbname="<data;base>"'

If no valid connection is found, the tests will use the "initdb" 
program to try and create a Postgres database cluster to test with.
The first available port starting at 5440 will be used.

You can increase the verbosity of the tests by setting the 
environment variable TEST_VERBOSE. You can also enable tracing 
within the tests themselves by setting DBD_TRACE to whatever 
trace level you want. Be aware that setting the trace level can 
result in extremely verbose output.

When reporting test failures, please use TEST_VERBOSE=1, but do *not* 
set DBD_TRACE unless requested, and send only the relevant sections.

If the first connection fails and DBI_DSN is not set, the tests may 
attempt to connect as the user 'postgres' to the database 'postgres'. 
See the t/dbdpg_test_setup.pl for the exact process.

Please consider installing CPAN::Reporter so that your tests are 
automatically gathered and reported, which helps the development 
of DBD::Pg.

TROUBLESHOOTING:
----------------

* Placeholder issues

If you find that some of your queries containing placeholders are no 
longer working, this may because DBD::Pg now uses the native PostgreSQL 
placeholders on the server itself whenever possible. Previously, DBD::Pg 
did a simple emulation of placeholders, so the rules were not as strict.
You should either rewrite your queries to make them legal SQL syntax for 
PostgreSQL, or turn off server-side prepares.

To change your queries, make sure that the type of each placeholder can 
be determined by the PostgreSQL parser. So instead of:

  SELECT ?

use something like:

  SELECT ?::int

To turn off server-side prepares completely (with a loss of some performance 
and features), do this at the top of your scripts:

$dbh->{pg_server_prepare} = 0;

This can also be set for individual queries at the statement handle level: see 
the documentation for more details.

* PostgreSQL library issues:

DBD::Pg uses the libpq library that comes with Postgres. If the shared libpq 
library is not available, DBD::Pg will error with a message that 
usually mentions a file names libpq.so, like this:

    Can't load './blib/arch/auto/DBD/Pg/Pg.so' for module DBD::Pg: libpq.so.5: cannot open 
    shared object file: No such file or directory at .../DynaLoader.pm line 230.

This means that the libraries are not installed in a place where the system 
can find them when it tries to load the Pg.so file. On some systems, you 
can run /sbin/ldconfig -v to see a list of shared modules, or just search 
the system for the file with "locate libpq.so". If it exists but is not being 
loaded, you may need to add the directory it is in to /etc/ld.so.conf file 
and run the ldconfig command. Otherwise, you may need to add the path to 
the environment variable LD_LIBRARY_PATH.

If you get an error message like:

	perl: error while loading shared libraries:
	/usr/lib/perl5/site_perl/5.6.1/i386-linux/auto/DBD/Pg/Pg.so: undefined
	symbol: PQconnectdb

when you call DBI->connect, then your libpq.so was probably not seen at
build-time.  This should have caused 'make test' to fail; did you really
run it and look at the output? 

* Perl issues:

Some Linux distributions have incomplete perl installations.  If you have
compile errors like "XS_VERSION_BOOTCHECK undeclared", do:

	find .../lib/perl5 -name XSUB.h -print

If this file is not present, you need to recompile and re-install perl.


If you get a message about "use of uninitialized value in -d" when doing 
a "make install_vendor", you can work around this by adding a dummy value 
to the INSTALLVENDORBIN environment variable:

make install_vendor INSTALLVENDORBIN=/tmp
(thanks to Peter Eisentraut <peter_e at gmx.net>)


* SGI issues:

If you get segmentation faults, make sure you are using the malloc
which comes with perl when compiling perl (the default is not to).
(thanks to "David R. Noble" <drnoble at engsci.sandia.gov>)


* HP issues:

If you get error messages like:

	can't open shared library: .../lib/libpq.sl
	No such file or directory

when running the test script, try to replace the 'shared' option in the
LDDFLAGS with 'archive'.
(thanks to Dan Lauterbach <danla at dimensional.com>)


* FreeBSD issues:

If you get during "make test" the error message:

	'DBD driver has not implemented the AutoCommit attribute'

recompile the DBI module and the DBD-Pg module and disable optimization.
This error message is due to the broken optimization in gcc-2.7.2.1.


If you get compiler errors like:

	In function `XS_DBD__Pg__dr_discon_all_'
	`sv_yes' undeclared (first use in this function)

it may be because there is a 'patchlevel.h' file from another package 
(such as 'hdf') in your POSTGRES_INCLUDE dir.  The presence of this file 
prevents the compiler from finding the perl include file 
'mach/CORE/patchlevel.h'.  Do 'pg_config --includedir' to identify the 
POSTGRES_INCLUDE dir.  Rename patchlevel.h whilst you build DBD::Pg. 


* Sun issues:

If you get compile errors like:

	/usr/include/string.h:57: parse error before `]'

then you need to remove from pgsql/include/libpq-fe.h the define for
strerror, which clashes with the definition in the standard include
file.


* Win32 issues:

For installation, please see the README.win32 file.

Running DBD-Pg scripts on Win32 needs some configuration work
on the server side:

    o add a postgres user with the same name as the NT-User 
      (e.g. Administrator)
    o make sure, that your pg_hba.conf on the server is configured,
      such that a connection from another host will be accepted


* OS X issues:

You may need to add "-lssl" and "-lcrypto" to your LIB variable 
before compiling.
(thanks to <rob at cabrion dot com>)

If having problems compiling, try running:

env -i command

This trick stops 'command' from inheriting environment variables from 
the shell process, which more often than not fixes up such weird build 
errors without having to do anything else in particular.
(thanks to David Landgren <david at landgren dot net>)


* SCO issues:

If the 'make test' gives an error about a symbol not being found, 
you can correct the problem by manually running ld after the 
'make' command:

LD_RUN_PATH="/usr/local/pgsql/lib" ld -G -L/usr/local/lib Pg.o \
dbdimp.o -o blib/arch/auto/DBD/Pg/Pg.so -L/usr/local/pgsql/lib -lpq \
-L/opt/K/SKUNK2000/Gcc/2.95.2pl1/usr/local/lib/gcc-lib/i386-pc-sco3.2v5.0.5/2.95.2/ \
-lgcc

Once this is done, 'make test' succeeds properly.
(thanks to <jmore at remote-print.com>)

COPYRIGHT:
----------

	Copyright (c) 2002-2008 Greg Sabino Mullane and others: see the Changes file
	Portions Copyright (c) 2002 Jeffrey W. Baker
	Portions Copyright (c) 1997-2001 Edmund Mergl
	Portions Copyright (c) 1994-1997 Tim Bunce

You may distribute under the terms of either the GNU General Public
License or the Artistic License, as specified in the Perl README file.



Changes
('GSM' is Greg Sabino Mullane, greg@turnstep.com)

2.8.8 Released December 17, 2009

  - Security release to fix high bit character problem in bytea
    (CPAN bug #51153) (Debian bug #554489)

2.8.7 Released July 24, 2008 (subversion r11582)

  - Modify test scripts to work better on FreeBSD boxes. [GSM]
  - Much documentation improvement and POD tweaking. [GSM]

2.8.6 Released July 21, 2008 (subversion r11558)

  - More testing improvements to increase odds of all tests being
    run, especially when testing as root. [GSM]

2.8.5 Released July 13, 2008 (subversion r11529)

  - Fix an obscure bug in which a coredump occurs if client_min_messages 
    is set to DEBUG3 or greater, and we then exit without disconnecting 
    while AutoCommit is off. The new behavior is to simply not attempt to 
    output the debugging information about the final 'rollback'. [GSM]
  - More documentation improvements. [GSM]

2.8.4 Released July 10, 2008 (subversion r11520)

  - Minor Perl::Critic test adjustments. [GSM]
  - Documentation enhancements. [GSM]
  - Yet more minor testing tweaks. [GSM]

2.8.3 Released July 6, 2008 (subversion r11495)

  - Minor testing functionality tweaks, lots of test cleanups, minor doc enhancements. [GSM]

2.8.2 Released June 29, 2008 (subversion r11466)

  - Minor testing tweaks, doc fixes. [GSM]

2.8.1 Released June 11, 2008 (subversion r11417)

  - Force testing to use a custom socket dir, to avoid 
    permission problems. Thanks to Frank Wiegand for 
    help in uncovering this. [GSM]

2.8.0 Released June 1, 2008 (subversion r11366)

  - Added in payload strings for LISTEN/NOTIFY in 8.4 
    via $dbh->pg_notifies() [GSM]
  - Fixed problem preventing some pg_type bind_arrays
    from working [GSM]
  - Fix tests in t.04misc.t to handle Windows newlines.
    (CPAN bug #36237) [Ian Macdonald]
  - Clean up get_info() information. [GSM]

2.7.2 Released May 14, 2008 (subversion r11269)

  - Handle embedded commas in quotes properly when destringifying 
    arrays. (CPAN bug #35862) [GSM]
  - Fix typo in docs with trace_parser_flags() (Martin J. Evans)
  - More testing tweaks [GSM]

2.7.1 Released May 11, 2008 (subversion r11250)

  - Yet more minor testing tweaks. [GSM]

2.7.0 Released May 10, 2008 (subversion r11239)

  - Have $dbh->quote() return E'' when server is >= 8.1 and string contains 
    backslashes. Fixes any problems with standard_conforming_strings.
    (CPAN bug #27538) [GSM]

2.6.6 Released May 7, 2008 (subversion r11214)

  - Fix minor problem in t/99_spellcheck.t [GSM]

2.6.5 Released May 7, 2008 (subversion r11209)

  - Add spell checker to tests. [GSM]
  - More tweaks to the testing suite. [GSM]

2.6.4 Released May 2, 2008 (subversion r11186)

  - More tweaks to the test suite. [GSM]

2.6.3 Released May 1, 2008 (subversion r11169)

  - Minor tweaks to the test suite. [GSM]

2.6.2 Released April 30, 2008 (subversion r11161)

  - Fix coredump when pg_getcopydata copies 0 rows into a 
    freshly created var. (CPAN bug #35556) [David Harris]
  - Allow 'make test' create a test database from scratch if
    it cannot find an existing one to use. [GSM]

2.6.1 Released April 22, 2008 (subversion r11133)

  - Don't free placeholder section, fixes problem when using 
    more than one named placeholder with the same name.
    (CPAN bug #35303) [GSM]

2.6.0 Released April 16, 2008 (subversion r11095)

  - Make pg_notifies a true function, so that you can now
    use $dbh->pg_notifies instead of $dbh->func('pg_notifies') [GSM]
  - Various performance improvements [GSM]
  - Fix minor build and compilation issues with Strawberry Perl [GSM]
  - Add Bundle::DBD::Pg [GSM]

2.5.1 Released April 7, 2008 (subversion r11056)

  - Correctly handle negative PID numbers on Win32 systems when
    generating prepared statement names (CPAN bug # 34738) [GSM]

2.5.0 Released March 23, 2008 (subversion r10990)

  - Add pg_enum_values to $dbh->column_info()
    [Dave Rolsky] (CPAN ticket #34351)
  - Minor test fixes. [GSM]

2.4.0 Released March 21, 2008 (subversion r10974)

  - Remove problematic and unneeded Test::Warn test from 00basic.t.
  - Add $sth->{pg_current_row} [GSM]

2.3.0 Released March 19, 2008 (subversion r10951)

  - Add $sth->{pg_bound} and $sth->{pg_numbound} [GSM]
  - Fix broken call to $sth->{pg_segments} [GSM]

2.2.2 Released March 3, 2008 (subversion r10873)

  - Remove non-working tracing from types.c and quote.c [GSM]
  - Add parse_trace_flag as statement handle method. [GSM]

2.2.1 Released March 1, 2008 (subversion r10865)

  - Fix memory leaks in dbdimp.c (CPAN bug #33743) [Alexey Tourbin]
  - Fix strlen problems in dbdimp.c (CPAN bug #33737) [Alexey Tourbin]
  - Fix char count in Renew() (CPAN bug #33738) [Alexey Tourbin]
  - Change local trace_flags to lowercase. [GSM]

2.2.0 Released February 27, 2008 (subversion r10849)

  - Introduce enhanced trace flags. See the documentation 
     on parse_trace_flags() for details. [GSM]
  - Remove version.pm dependency from Makefile.PL (CPAN bug #33429)

2.1.3 Released February 20, 2008 (subversion r10813)

  - Do not assume POSTGRES_LIB is a plain dirname, as it may have " -lssl".

2.1.2 Released February 19, 2008 (subversion r10806)

  - Do not build if environment variables POSTGRES_HOME, POSTGRES_LIB, 
    or POSTGRES_INCLUDE are set but not valid.
  - Fix dependency requirements, especially version.pm [GSM]

2.1.1 Released February 19, 2008 (subversion r10798)

  - Better URLs to cpan.org resources. [GSM]

2.1.0 Released February 18, 2008 (subversion r10793)

  - Use version.pm (CPAN bug #33206) [GSM]
  - Add PERL_NO_GET_CONTEXT #define to improve performance on threaded Perls [GSM]
  - Raise the minimum DBI version to 1.52. [GSM]
  - Allow arrayrefs into bind_col (CPAN bug #33193) [GSM]
  - Remove '//' style comments to make strict ANSI compilers happy.
    [Trevor Inman] (CPAN bug #33089)
  - Force client encoding of UTF8 for some tests. [GSM]

Makefile.PL
# $Id: Makefile.PL 13673 2009-12-17 17:06:50Z turnstep $

use ExtUtils::MakeMaker;
use Config;
use strict;
use warnings;
use 5.006001;

## No version.pm for this one, as the prereqs are not loaded yet.
my $VERSION = '2.16.0';

my $lib;
BEGIN {
	use vars qw/$sep/;
    my %sep = (
			   MacOS   => ':',
               MSWin32 => '\\',
               os2     => '\\',
               VMS     => '\\',
               NetWare => '\\',
               dos     => '\\',
			   );
    $sep = $sep{$^O} || '/';
    $lib = join $sep, 't', 'lib';
}

use lib $lib;
if ($VERSION =~ /_/) {
	print "WARNING! This is a test version ($VERSION) and should not be used in production!\n";
}

if (grep { /help/ } @ARGV) {
	print qq{
Usage: perl $0

No other options are necessary, although you may need to
set some evironment variables. See the README file for full details.

In brief:

By default Makefile.PL uses App::Info to find the location of the
PostgreSQL library and include directories.  However, if you want to
control it yourself, define the environment variables POSTGRES_INCLUDE
and POSTGRES_LIB, or define just POSTGRES_HOME. Note that if you have
compiled PostgreSQL with SSL support, you must define the POSTGRES_LIB
environment variable and add "-lssl" to it, like this:

export POSTGRES_LIB="/usr/local/pgsql/lib -lssl"

The usual steps to install DBD::Pg:

1.   perl Makefile.PL
2.   make
3.   make test
4.   make install

Do steps 1 to 3 as a normal user, not as root!

If all else fails, email dbd-pg\@perl.org for help.

};
	exit 1;

}

print "Configuring DBD::Pg $VERSION\n";

my $POSTGRES_INCLUDE;
my $POSTGRES_LIB;

# We need the version information to properly set compiler options later
# Use App::Info to get the data we need.
require App::Info::RDBMS::PostgreSQL;
require App::Info::Handler::Prompt;
my $p = App::Info::Handler::Prompt->new;
my $pg = App::Info::RDBMS::PostgreSQL->new(on_unknown => $p);
my ($major_ver, $minor_ver, $patch, $conf, $bindir) = map {$pg->$_}
	qw/major_version minor_version patch_version configure bin_dir/;
my $initdb = '';
if (defined $bindir and -d $bindir) {
	my $testinitdb = "$bindir${sep}initdb";
	if (-e $testinitdb) {
		$initdb = $testinitdb;
	}
}
my $serverversion = 0;
my $defaultport = 0;

if (defined $major_ver) {
	$serverversion = sprintf '%d%.02d%.02d', $major_ver, $minor_ver, $patch;
	$defaultport = $conf =~ /with-pgport=(\d+)/ ? $1 : 5432;
}

# We set POSTGRES_INCLUDE and POSTGRES_LIB from the first found of:
# 1. environment variable
# 2. App::Info::RDBMS::PostgreSQL information
# 3. subdirectory of $ENV{POSTGRES_HOME}

$POSTGRES_INCLUDE = $ENV{POSTGRES_INCLUDE} || $pg->inc_dir || "$ENV{POSTGRES_HOME}/include";

$POSTGRES_LIB = $ENV{POSTGRES_LIB} || $pg->lib_dir || "$ENV{POSTGRES_HOME}/lib";

my $os = $^O;
print "PostgreSQL version: $serverversion (default port: $defaultport)\n";
my $showhome = $ENV{POSTGRES_HOME} || '(not set)';
print "POSTGRES_HOME: $showhome\n";
my $showinc = $POSTGRES_INCLUDE || '(not set)';
print "POSTGRES_INCLUDE: $showinc\n";
my $showlib = $POSTGRES_LIB || '(not set)';
print "POSTGRES_LIB: $showlib\n";
print "OS: $os\n";

my $baddir = 0;
sub does_path_exist {
	my ($path_name, $path) = @_;

	return if ! defined $path or ! length $path or -d $path;
	printf "The value of %s points to a non-existent directory: %s\n",
		$path_name, $path;
	$baddir++;
	return;
}

does_path_exist('POSTGRES_HOME', $ENV{POSTGRES_HOME});
does_path_exist('POSTGRES_INCLUDE',  $POSTGRES_INCLUDE);

if ($baddir) {
	print "Cannot build unless the directories exist, exiting.\n";
	exit 0;
}

if ($serverversion < 11) {
	print "Could not determine the PostgreSQL library version.\n".
		"Please ensure that a valid path is given to the 'pg_config' command,\n".
			"either manually or by setting the environment variables\n".
				"POSTGRES_DATA, POSTGRES_INCLUDE, and POSTGRES_LIB\n";
	exit 0;
}

if ($os =~ /Win32/) {
	for ($POSTGRES_INCLUDE, $POSTGRES_LIB) {
		$_ = qq{"$_"} if index $_,'"';
	}
}

## Warn about older versions
if ($serverversion < 70400) {
	print "\n****************\n";
	print "WARNING! DBD::Pg no longer supports versions less than 7.4.\n";
	print "You must upgrade PostgreSQL to a newer version.\n";
	print "****************\n\n";
	exit 1;
}

my $dbi_arch_dir;
{
	eval {
		require DBI::DBD;
	};
	if ($@) {
		print "Could not load DBI::DBD - is the DBI module installed?\n";
		exit 0;
	}
	local *STDOUT; ## Prevent duplicate debug info as WriteMakefile also calls this
	$dbi_arch_dir = DBI::DBD::dbd_dbi_arch_dir();
}

my $defines = " -DPGLIBVERSION=$serverversion -DPGDEFPORT=$defaultport";
my $comp_opts = $Config{q{ccflags}} . $defines;

if ($ENV{DBDPG_GCCDEBUG}) {
	warn "Enabling many compiler options\n";
	$comp_opts .= ' -Wchar-subscripts -Wcomment';
	$comp_opts .= ' -Wformat=2'; ## does -Wformat,-Wformat-y2k,-Wformat-nonliteral,-Wformat-security
	$comp_opts .= ' -Wnonnull';
	$comp_opts .= ' -Wuninitialized -Winit-self'; ## latter requires the former
	$comp_opts .= ' -Wimplicit'; ## does -Wimplicit-int and -Wimplicit-function-declaration
	$comp_opts .= ' -Wmain -Wmissing-braces -Wparentheses -Wsequence-point -Wreturn-type -Wswitch -Wswitch-enum -Wtrigraphs';
	$comp_opts .= ' -Wunused'; ## contains -Wunused- function,label,parameter,variable,value
	$comp_opts .= ' -Wunknown-pragmas -Wstrict-aliasing';
	$comp_opts .= ' -Wall'; ## all of above, but we enumerate anyway
	$comp_opts .= ' -Wextra -Wdeclaration-after-statement -Wendif-labels -Wpointer-arith';
	$comp_opts .= ' -Wbad-function-cast -Wcast-qual -Wcast-align -Wsign-compare -Waggregate-return';
	$comp_opts .= ' -Wmissing-prototypes -Wmissing-declarations -Wmissing-format-attribute -Wpacked -Winline -Winvalid-pch';
	$comp_opts .= ' -Wdisabled-optimization';
	$comp_opts .= ' -Wnested-externs';
	$comp_opts .= " -Wstrict-prototypes"; ## Still hits a couple places in types.h
	$comp_opts .= " -Wswitch-default";
	$comp_opts .= " -Wsystem-headers";
	$comp_opts .= " -Wmissing-noreturn";
	$comp_opts .= " -Wfloat-equal"; ## Does not like SvTRUE() calls
	$comp_opts .= " -Wpadded"; ## Use when adding/changing our structs
}

my %opts =
	(
	 NAME           => 'DBD::Pg',
	 VERSION_FROM   => 'Pg.pm',
	 INC            => "-I$POSTGRES_INCLUDE -I$dbi_arch_dir",
	 OBJECT         => "Pg\$(OBJ_EXT) dbdimp\$(OBJ_EXT) quote\$(OBJ_EXT) types\$(OBJ_EXT)",
	 LIBS           => ["-L$POSTGRES_LIB -lpq -lm"],
	 AUTHOR         => 'Greg Sabino Mullane',
	 ABSTRACT       => 'PostgreSQL database driver for the DBI module',
	 PREREQ_PM      => {
						'ExtUtils::MakeMaker' => '6.11',
						'DBI'                 => '1.52',
						'Test::More'          => '0.61',
						'version'             => '0',
					   },
	 CCFLAGS        => $comp_opts,
	 PERL_MALLOC_OK => 1,
	 NEEDS_LINKING  => 1,
	 NO_META        => 1,
     NORECURS       => 1,
	 clean          => { FILES => 'trace Pg.xsi README.testdatabase' },
     realclean      => { FILES => 'dbdpg_test_database/' },
);

if ($os eq 'hpux') {
    my $osvers = $Config{osvers};
    if ($osvers < 10) {
        print "Warning: Forced to build static not dynamic on $os $osvers.\a\n";
        $opts{LINKTYPE} = 'static';
    }
}
elsif ($os =~ /Win32/) {
	my $msdir = $POSTGRES_LIB;
	$msdir =~ s{"$}{/ms"};
	$opts{LIBS}[0] .= " -L$msdir -lsecur32";
}

if ($Config{dlsrc} =~ /dl_none/) {
    $opts{LINKTYPE} = 'static';
}

{
	package MY; ## no critic
	sub MY::test { ## no critic
		my $string = shift->SUPER::test(@_);
		$string =~ s/(PERL_DL_NONLAZY=1)/PGINITDB="$initdb" $1/g;
		return $string;
	}
}

sub constants {
	my $self = shift;

	my $old_constants = $self->SUPER::constants();
	my $new_constants = '';
	for my $line (split /\n/ => $old_constants) {
		if ($line =~ /^INC = .*strawberry.*/ ) {
			print qq(Strawberry Perl found; adjusting the INC variable;\n);
			$line . ' -I ' . DBI::DBD::dbd_dbi_arch_dir();
			print qq(INC is now $line\n);
		}
		$new_constants .= "$line\n";
	}
	return $new_constants;
}

sub MY::postamble { ## no critic ProhibitQualifiedSubDeclarations
	no strict 'subs'; ## no critic ProhibitNoStrict
	my $string = DBI::DBD->dbd_postamble();
	use strict 'subs';
	## Evil, evil stuff - but we really want to suppress the "duplicate function" message!
	$string =~ s/dependancy/dependency/g; ## why not, while we are here
	$string =~ s{(BASEEXT\)/g)}{$1; s/^do\\\(/dontdo\\\(/};

        my $tags = <<'MAKE_FRAG';
.PHONY: tags

tags:
	ctags -f tags --recurse --totals \
		--exclude=blib \
		--exclude=.svn \
		--exclude='*~' \
		--languages=Perl,C --langmap=c:+.h,Perl:+.t \

MAKE_FRAG
        $string = "$string\n$tags\n";

	$string .= <<'MAKE_SPLINT';

## This must be version 3.1.2 or better: earlier versions have many 
## problems parsing the DBI header files
SPLINT = splint

## Temp directory, for use with +keep
SPLINT_TMP = $(TMP)/splint_dbdpg

SPLINTFLAGS =            \
  -message-stream-stdout \
  -linelen 90            \
  -boolops               \
  -tmpdir $(SPLINT_TMP)  \
  +posixstrictlib        \
  +ignoresigns           \
  +showdeephistory       \
  -predboolint           \
  -nullpass              \
  +charint               \
  +boolint               \
  +allglobals            \

SPLINTFLAGS_TEST = 

SDEFINES = 

CCFLAGS2 = -D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DPGLIBVERSION=80111 -DPGDEFPORT=5432

splint: $(H_FILES) $(C_FILES)
	$(MKPATH) $(SPLINT_TMP)
	$(SPLINT) $(SPLINTFLAGS) $(SPLINTFLAGS_TEST) $(SDEFINES) -I$(PERL_INC) $(INC) $(C_FILES)

MAKE_SPLINT

	$string =~ s/SDEFINES = /SDEFINES =$defines/;

	return $string;
}

my $output = WriteMakefile(%opts);

if (!exists $output->{EXTRALIBS} or

	($output->{EXTRALIBS} !~ /\-lpq/ and $output->{EXTRALIBS} !~ /libpq/)) {

	my $makefile = exists $output->{MAKEFILE}
		? "\nRemoving ($output->{MAKEFILE})\n" : '';

	warn qq{
==========================================================

WARNING! No libpq libraries were detected!

You need to install the postgresql-libs package for your system,

or set the POSTGRES_LIB environment variable to the correct place.
$makefile
===========================================================

};

	## Do not let make proceed
	unlink $output->{MAKEFILE} if $makefile;

	exit 1;
}

exit 0;

# end of Makefile.PL
